File: server/balu/sync/recurrence.py
"""RRULE-subset parser and next-occurrence calculation.

Supported grammar (contract §3.3):
    FREQ=DAILY|WEEKLY|MONTHLY|YEARLY [;INTERVAL=n] [;BYDAY=MO,TU,WE,TH,FR,SA,SU]
BYDAY is only valid with FREQ=WEEKLY. Week starts Monday (WKST=MO).
"""

from __future__ import annotations

import calendar
from dataclasses import dataclass, field
from datetime import date, timedelta

_WEEKDAYS = {"MO": 0, "TU": 1, "WE": 2, "TH": 3, "FR": 4, "SA": 5, "SU": 6}
_FREQS = {"DAILY", "WEEKLY", "MONTHLY", "YEARLY"}


@dataclass
class Recurrence:
    freq: str
    interval: int = 1
    byday: list[int] = field(default_factory=list)


def parse_recurrence(rrule: str) -> Recurrence:
    """Parse and validate an RRULE-subset string. Raises ValueError on any problem."""
    if not rrule or not isinstance(rrule, str):
        raise ValueError("empty recurrence")

    parts: dict[str, str] = {}
    for chunk in rrule.strip().split(";"):
        if not chunk:
            continue
        if "=" not in chunk:
            raise ValueError(f"malformed component: {chunk!r}")
        key, _, value = chunk.partition("=")
        parts[key.strip().upper()] = value.strip().upper()

    freq = parts.get("FREQ")
    if freq not in _FREQS:
        raise ValueError(f"invalid FREQ: {freq!r}")

    interval = 1
    if "INTERVAL" in parts:
        try:
            interval = int(parts["INTERVAL"])
        except ValueError as exc:
            raise ValueError("INTERVAL must be an integer") from exc
        if interval < 1:
            raise ValueError("INTERVAL must be >= 1")

    byday: list[int] = []
    if "BYDAY" in parts:
        if freq != "WEEKLY":
            raise ValueError("BYDAY is only valid with FREQ=WEEKLY")
        for code in parts["BYDAY"].split(","):
            code = code.strip()
            if code not in _WEEKDAYS:
                raise ValueError(f"invalid BYDAY code: {code!r}")
            byday.append(_WEEKDAYS[code])
        byday = sorted(set(byday))

    unknown = set(parts) - {"FREQ", "INTERVAL", "BYDAY"}
    if unknown:
        raise ValueError(f"unsupported components: {sorted(unknown)}")

    return Recurrence(freq=freq, interval=interval, byday=byday)


def _add_months(d: date, months: int) -> date:
    total = d.month - 1 + months
    year = d.year + total // 12
    month = total % 12 + 1
    last_day = calendar.monthrange(year, month)[1]
    return date(year, month, min(d.day, last_day))


def _add_years(d: date, years: int) -> date:
    year = d.year + years
    last_day = calendar.monthrange(year, d.month)[1]
    return date(year, d.month, min(d.day, last_day))


def _nth(anchor: date, rec: Recurrence, k: int) -> date:
    """The k-th member of the series, always measured from `anchor`.

    Measuring from the anchor (rather than from the previous occurrence) is what
    keeps month-end rules stable: Jan 31 → Feb 28 → Mar 31, not Mar 28.
    """
    if rec.freq == "DAILY":
        return anchor + timedelta(days=rec.interval * k)
    if rec.freq == "WEEKLY":
        return anchor + timedelta(weeks=rec.interval * k)
    if rec.freq == "MONTHLY":
        return _add_months(anchor, rec.interval * k)
    if rec.freq == "YEARLY":
        return _add_years(anchor, rec.interval * k)
    raise ValueError(f"unhandled FREQ: {rec.freq}")  # pragma: no cover


def _start_k(anchor: date, rec: Recurrence, after: date) -> int:
    """A lower bound for the answer's index — never overshoots it."""
    if after <= anchor:
        return 0
    if rec.freq == "DAILY":
        return max(0, (after - anchor).days // rec.interval)
    if rec.freq == "WEEKLY":
        return max(0, (after - anchor).days // (7 * rec.interval))
    if rec.freq == "MONTHLY":
        months = (after.year - anchor.year) * 12 + (after.month - anchor.month)
        return max(0, months // rec.interval)
    return max(0, (after.year - anchor.year) // rec.interval)


_SCAN_LIMIT = 64


def next_occurrence(rrule: str, anchor: date, after: date | None = None) -> date:
    """The next occurrence of the series anchored at `anchor`, strictly after `after`.

    The series is `anchor, anchor+interval, anchor+2·interval, …` — for
    `FREQ=WEEKLY` with `BYDAY`, the BYDAY days of every `interval`-th week
    starting from the anchor's own week.

    Deriving everything from `after` instead (the pre-v1.2.1 behaviour) lost the
    rule's phase: every `INTERVAL > 1` rule then disagreed with the clients, and
    MONTHLY/YEARLY lost the original day-of-month. `after` defaults to `anchor`.
    Mirrored byte-for-byte by `packages/sync-client/src/recurrence.ts`.
    """
    rec = parse_recurrence(rrule)
    if after is None:
        after = anchor

    if rec.freq == "WEEKLY" and rec.byday:
        monday = anchor - timedelta(days=anchor.weekday())
        span = (after - monday).days
        first = max(0, span // (7 * rec.interval)) if span > 0 else 0
        for k in range(first, first + _SCAN_LIMIT):
            week = monday + timedelta(weeks=rec.interval * k)
            for weekday in rec.byday:  # sorted by parse_recurrence
                candidate = week + timedelta(days=weekday)
                if candidate > after:
                    return candidate
        raise ValueError("could not find the next occurrence")  # pragma: no cover

    first = _start_k(anchor, rec, after)
    for k in range(first, first + _SCAN_LIMIT):
        candidate = _nth(anchor, rec, k)
        if candidate > after:
            return candidate
    raise ValueError("could not find the next occurrence")  # pragma: no cover

File: packages/sync-client/src/recurrence.ts
// Client-side next-occurrence for the RRULE subset (contract §3.3). Used by the
// optimistic apply of `task_complete` on recurring tasks; the server remains
// authoritative — and must agree, so this file mirrors
// `server/balu/sync/recurrence.py` step for step.

import { addDaysISO, addMonthsISO, addYearsISO, diffDaysISO, dowISO } from "@balu/domain";

/** BYDAY code → offset from Monday (WKST=MO), matching the server's _WEEKDAYS. */
const BYDAY_TO_OFFSET: Record<string, number> = {
  MO: 0, TU: 1, WE: 2, TH: 3, FR: 4, SA: 5, SU: 6,
};

interface Rule {
  freq: "DAILY" | "WEEKLY" | "MONTHLY" | "YEARLY";
  interval: number;
  /** Offsets from the week's Monday, ascending. */
  byday: number[];
}

export function parseRrule(rrule: string): Rule | null {
  const parts = rrule.split(";");
  let freq: Rule["freq"] | null = null;
  let interval = 1;
  let byday: number[] = [];
  for (const p of parts) {
    const [k, v] = p.split("=");
    if (k === "FREQ" && (v === "DAILY" || v === "WEEKLY" || v === "MONTHLY" || v === "YEARLY")) {
      freq = v;
    } else if (k === "INTERVAL") {
      interval = Math.max(1, Number(v) || 1);
    } else if (k === "BYDAY") {
      byday = v.split(",").map((d) => BYDAY_TO_OFFSET[d]).filter((n) => n != null);
    }
  }
  if (!freq) return null;
  // BYDAY is only meaningful with FREQ=WEEKLY (§3.3), and the server iterates it
  // in ascending order — match both so the two engines cannot diverge.
  if (freq !== "WEEKLY") byday = [];
  byday = [...new Set(byday)].sort((a, b) => a - b);
  return { freq, interval, byday };
}

/** The Monday of `iso`'s week (WKST=MO). */
function mondayOf(iso: string): string {
  return addDaysISO(iso, -((dowISO(iso) + 6) % 7));
}

function year(iso: string): number {
  return Number(iso.slice(0, 4));
}
function month(iso: string): number {
  return Number(iso.slice(5, 7));
}

/**
 * The k-th member of the series, always measured from `anchor`. Measuring from
 * the anchor (rather than from the previous occurrence) is what keeps month-end
 * rules stable: Jan 31 → Feb 28 → Mar 31, not Mar 28.
 */
function nth(anchor: string, rule: Rule, k: number): string {
  switch (rule.freq) {
    case "DAILY":
      return addDaysISO(anchor, rule.interval * k);
    case "WEEKLY":
      return addDaysISO(anchor, 7 * rule.interval * k);
    case "MONTHLY":
      return addMonthsISO(anchor, rule.interval * k);
    case "YEARLY":
      return addYearsISO(anchor, rule.interval * k);
  }
}

/** A lower bound for the answer's index — never overshoots it. */
function startK(anchor: string, rule: Rule, after: string): number {
  if (after <= anchor) return 0;
  switch (rule.freq) {
    case "DAILY":
      return Math.max(0, Math.floor(diffDaysISO(anchor, after) / rule.interval));
    case "WEEKLY":
      return Math.max(0, Math.floor(diffDaysISO(anchor, after) / (7 * rule.interval)));
    case "MONTHLY": {
      const months = (year(after) - year(anchor)) * 12 + (month(after) - month(anchor));
      return Math.max(0, Math.floor(months / rule.interval));
    }
    case "YEARLY":
      return Math.max(0, Math.floor((year(after) - year(anchor)) / rule.interval));
  }
}

const SCAN_LIMIT = 64;

/**
 * The next occurrence of the series anchored at `anchor`, strictly after
 * `after`. The series is `anchor, anchor+interval, anchor+2·interval, …` — for
 * `FREQ=WEEKLY` with `BYDAY`, the BYDAY days of every `interval`-th week
 * starting from the anchor's own week.
 *
 * Returns an ISO date, or null if the rrule is unsupported.
 */
export function nextOccurrence(rrule: string, anchor: string, after: string): string | null {
  const rule = parseRrule(rrule);
  if (!rule) return null;

  if (rule.freq === "WEEKLY" && rule.byday.length > 0) {
    const monday = mondayOf(anchor);
    const span = diffDaysISO(monday, after);
    const first = span > 0 ? Math.max(0, Math.floor(span / (7 * rule.interval))) : 0;
    for (let k = first; k < first + SCAN_LIMIT; k += 1) {
      const week = addDaysISO(monday, 7 * rule.interval * k);
      for (const weekday of rule.byday) {
        const candidate = addDaysISO(week, weekday);
        if (candidate > after) return candidate;
      }
    }
    return null;
  }

  const first = startK(anchor, rule, after);
  for (let k = first; k < first + SCAN_LIMIT; k += 1) {
    const candidate = nth(anchor, rule, k);
    if (candidate > after) return candidate;
  }
  return null;
}
